### DISCUSSION TITLE
Using a pre-trained model from another framework as base

### DISCUSSION THREAD

**Discussion by ganesh-krishnan** [online-ml/river#850] [Q&A] [answered]
I'm interested in creating a structure where I can add online learning capabilities to any existing model (EM for short). The EM could have been trained in any framework (eg. `scikit-learn`, `keras` etc.). I took a look at this [discussion](https://github.com/online-ml/river/discussions/478). But it assumes that the EM is a `river` model. What's the best way to use some other model as a base? One idea I had was to train a `river` model on the residual errors from the EM in online fashion. Don't know if there are other better ideas.

**Comment by gbolmier** [comment, marked as answer]
Hey @ganesh-krishnan, sorry for the slow reply. It's actually pretty use-case dependent, I don't see a super generic way here when implementations don't follow common standards.

When transferring a model state from a framework to another, implementation details from both frameworks matter a lot. There is some leg work to do to first to check if implementations are compatible and how to make the conversion. Transformers of your training pipeline matter too. If you have a standard scaler for instance, you probably want to transfer its state too.

If you're willing to describe more the context and participate in the solution implementation we could definitely give a hand. If this ends up in a clean solution for your use-case, I think it would make sense to be added in [river-extra](https://github.com/online-ml/river-extra) as a first step.

**Comment by ganesh-krishnan** [comment]
@gbolmier Thank you for the response. We are talking about two different things I think. I'm not talking about transferring state which like you pointed out will be very specific to individual frameworks. I'm talking about something like this:

1. Train an "offline" model in framework of choice to predict quantity of interest
2. Initialize an online river model to learn prediction errors
3. At inference time, prediction = prediction from offline model + error prediction from river model
4. Update online model (river model of prediction errors) when new data comes in ie. calculate the target as the prediction error between actuals and the offline model 

Let me know if that makes sense. This is of course, one way of adding online learning capabilities to any black box offline model. Wondering if there are other ways.

**Reply by MaxHalford** [reply]
I've personally never heard of such an approach. It feels a bit convoluted and over-engineered. But I would simply suggest trying this out. Or do you have any specific question?

**Reply by gbolmier** [reply]
Same for me. Might introduced quite some overhead in your workflow but could also work in practice, depends on your context. Best is to give it a try.
